app/models/concerns/user_edits.rb

```diff
@@ -0,0 +1,45 @@
+module UserEdits
+  class Edits
+    PER_PAGE = 20
+    EDITABLE_TYPES = %w[Relationship Entity List Document].freeze
+    UserEdit = Struct.new(:resource, :version, :action, :time)
+    VersionsToModel = proc { |vs| vs.first.item_type.constantize.unscoped.find(vs.map(&:item_id).uniq) }
+
+    def initialize(user, page: 1)
+      @user = user
+      @page = page
+    end
+
+    def recent_edits
+      @recent_edits ||= PaperTrail::Version
+                          .where(whodunnit: @user.id, item_type: EDITABLE_TYPES)
+                          .order(id: :desc)
+                          .page(@page)
+                          .per(PER_PAGE)
+    end
+
+    def recent_edits_presenter
+      recent_edits.map do |v|
+        UserEdit.new(
+          ar_lookup.dig(v.item_type, v.item_id),
+          v,
+          v.event.tr('_', ' ').capitalize,
+          v.created_at.strftime('%B %e, %Y%l%p')
+        )
+      end
+    end
+
+    private
+
+    #  {
+    #    'Entity' => { 123 => <Entity>, '345'=> <Entity> },
+    #     Relationship' => { 456 => <Relationship> }
+    #  }
+    def ar_lookup
+      @ar_lookup ||= recent_edits
+                       .group_by { |v| v.item_type }
+                       .transform_values(&VersionsToModel)
+                       .transform_values { |models| models.map { |m| [m.id, m] }.to_h }
```

Comment:
mad nitpicky, but if you wind up writing this pattern a lot (it's a useful pattern!), it could *maybe* be terser if you either (1) do the `models.map` bit from the second `.transform_values` call as a `map` chained to the `VersionsToModel` call in the prior `transform_values` and/or (2) just wrote another proc called `ModelsToHash` that you call in a second (and reuse in such calls in the future) and/or (3) just stuffed all that logic directly into a `VersionsToModelHashes` proc? r

ealize it's just a bunch of shuffling of the same logic into different containers, but the space to take a breath and name what the last operation is doing with a function name and do one less block/pipe thingie might make a *slight* difference in readability.

like i said: mad nitpicky. this is great code. just riffing.

Reply:
thanks! that's a good idea. i made another proc for the second transform values: 7b91eb066c3d7da8fb84ace3a8c5aaf76f250015